src/main/scala/viper/gobra/ast/frontend/Ast.scala

```diff
@@ -416,12 +423,33 @@ case class PExpCompositeVal(exp: PExpression) extends PCompositeVal // exp is ne
 
 case class PLitCompositeVal(lit: PLiteralValue) extends PCompositeVal
 
-case class PFunctionLit(args: Vector[PParameter], result: PResult, body: PBlock) extends PLiteral with PCodeRootWithResult with PScope
+case class PFunctionLit(decl: PClosureNamedDecl) extends PLiteral with PActualExpression
+
+case class PClosureNamedDecl(id: Option[PIdnDef], decl: PClosureDecl) extends PGhostMisc {
+  val result: PResult = decl.result
+}
+
+case class PClosureDecl(args: Vector[PParameter],
+                  result: PResult,
+                  spec: PFunctionSpec,
+                  body: Option[(PBodyParameterInfo, PBlock)]) extends PFunctionOrClosureDecl with PCodeRootWithResult with PActualMisc
+
+case class PClosureSpecInstance(func: PNameOrDot, params: Vector[PClosureSpecParameter]) extends PGhostMisc
+
+case class PClosureSpecParameter(key: Option[PClosureSpecParameterKey], exp: PExpression) extends PGhostMisc
+
+case class PClosureSpecParameterKey(name: String) extends PGhostMisc
```

Comment:
Do we really need `PClosureSpecParameter` and `PClosureSpecParameterKey`? Couldn't we use `PKeyedElement` here instead? Or does it become cumbersome because we would need to deal with `PCompositeVal`s?

Reply:
I remember trying to use `PKeyedElement` and that caused some problems with name resolution, but I can't remember exactly so I will double check and see what happens if I try using that

Reply:
I tried; using `PKeyedElement` does cause problems with name resolution, but those could be solved relatively simply by adding a new rule in the `entity` attribute, so I opted for it.

So I have removed `PClosureSpecParameter` and `PClosureSpecParameterKey`